event.go

```diff
@@ -303,6 +313,117 @@ func (e Event) StateKeyEquals(stateKey string) bool {
 	return *e.fields.StateKey == stateKey
 }
 
+const (
+	// The event ID, room ID sender, event type and state key fields cannot be
+	// bigger than this.
+	// https://github.com/matrix-org/synapse/blob/v0.21.0/synapse/event_auth.py#L173-L182
+	maxIDLength = 255
+	// The entire event JSON, including signatures cannot be bigger than this.
+	// https://github.com/matrix-org/synapse/blob/v0.21.0/synapse/event_auth.py#L183-184
+	maxEventLength = 65536
+)
+
+// CheckFields checks that the event fields are valid.
+// Returns an error if the IDs have the wrong format or too long.
+// Returns an error if the total length of the event JSON is too long.
+// Returns an error if the event ID doesn't match the origin of the event.
+func (e Event) CheckFields() error {
+	if len(e.eventJSON) > maxEventLength {
+		return fmt.Errorf(
+			"gomatrixserverlib: event is too long, length %d > maximum %d",
+			len(e.eventJSON), maxEventLength,
+		)
+	}
+
+	if len(e.fields.Type) > maxIDLength {
+		return fmt.Errorf(
+			"gomatrixserverlib: event type is too long, length %d > maximum %d",
+			len(e.fields.Type), maxIDLength,
+		)
+	}
+
+	if e.fields.StateKey != nil && len(*e.fields.StateKey) > maxIDLength {
+		return fmt.Errorf(
+			"gomatrixserverlib: state key is too long, length %d > maximum %d",
+			len(*e.fields.StateKey), maxIDLength,
+		)
+	}
+
+	_, err := checkID(e.fields.RoomID, "room", '!')
+	if err != nil {
+		return err
+	}
+
+	senderDomain, err := checkID(e.fields.Sender, "user", '@')
+	if err != nil {
+		return err
+	}
+
+	eventDomain, err := checkID(e.fields.EventID, "event", '$')
```

Comment:
Hmm, should we be doing this? http://matrix.org/docs/spec/intro.html#room-ids-and-event-ids

> The domain is used only for namespacing to avoid the risk of clashes of identifiers between different homeservers.

But yet we assert here that the event is in this format?